Library/Homebrew/utils.rb

```diff
@@ -770,3 +770,39 @@ def number_readable(number)
   (numstr.size - 3).step(1, -3) { |i| numstr.insert(i, ",") }
   numstr
 end
+
+# True if this version of Ruby supports text encodings in its strings
+def ruby_has_encoding?
+  String.method_defined?(:force_encoding)
+end
+
+# Truncates a text string to fit within a byte size constraint,
+# preserving character encoding validity. The returned string will
+# be not much longer than the specified max_bytes, though the exact
+# shortfall or overrun may vary.
+def truncate_text_to_approximate_size(s, max_bytes, options = {})
+  front_weight = options.fetch(:front_weight, 0.5)
+  return s if s.bytesize <= max_bytes
+
+  if front_weight < 0.0 || front_weight > 1.0
+    raise "opts[:front_weight] must be between 0.0 and 1.0"
+  end
+  glue = "\n[...snip...]\n"
+  max_bytes_in = [max_bytes - glue.bytesize, 1].max
+  if ruby_has_encoding?
+    bytes = s.dup.force_encoding("BINARY")
+    glue_bytes = glue.encode("BINARY")
+  else
+    bytes = s
+    glue_bytes = glue
+  end
+  front = bytes[0..(max_bytes_in * front_weight).floor]
+  back = bytes[-(max_bytes_in * (1.0 - front_weight)).ceil..-1]
```

Comment:
This needs a `+ 1` added inside the parentheses:

```ruby
back = bytes[-(max_bytes_in * (1.0 - front_weight) + 1).ceil..-1]
```

Otherwise it will do the wrong thing for the case of `front_weight = 1.0` as this will result in a range `0..-1` and this will yield the complete string instead of the bare minimum (the last character). I believe there's no simple way to end up with an empty string for the back part for `front_weight = 1.0`. (I guess the same can be said about the front part and the special case `front_weight = 0.0`.)